// main/consts.js
/**
 * This module has the important constants which are used in the entire app - main and renderer threads -
 *
 * Node.js path module
 * Working with file and directory paths, and provide useful utilities
 */
const Path = require('path')

// Define all constants
const Constants = {
  // When set to `false`, the AI assistant will be completely hidden in the UI and no way to interact with it
  EnableAIAssistant: true,
  // The host server where the AI assistant Python project is running
  AIAssistantServer: 'http://johnnyadhoc.oberk123.com:1512',
  /**
   * A server that accepts the crash reports' POST requests
   * The crash reporter project is hosted on GitHub
   */
  CrashHandlerServer: 'http://127.0.0.1:55155/set/crash',
  // Default `cqlsh.rc` file's content
  CQLSHRC: require(Path.join(__dirname, '..', '..', 'renderer', 'js', 'default_cqlshrc.js')).CQLSHRC,
  // Default `docker-compose.yml` content
  DockerComposeYAML: require(Path.join(__dirname, '..', '..', 'renderer', 'js', 'default_docker_compose_yml.js')).DockerYAML,
  // Define sensitive data to be checked inside the `cqlsh.rc` content
  SensitiveData: ['username', 'password'],
  // Allowed HTML tags to be rendered in the localization process
  AllowedHTMLTags: ['b', 'code', 'kbd', 'br'],
  /**
   * Define CQL keywords
   *
   * From:
   * https://github.com/apache/cassandra/blob/trunk/src/resources/org/apache/cassandra/cql3/reserved_keywords.txt
   */
  CQLKeywords: ['ADD', 'ALLOW', 'ALTER', 'AND', 'APPLY', 'ASC', 'AUTHORIZE', 'BATCH', 'BEGIN', 'BY', 'COLUMNFAMILY', 'CREATE', 'DELETE', 'DESC', 'DESCRIBE', 'DROP', 'ENTRIES', 'EXECUTE', 'FROM', 'FULL', 'GRANT', 'IF', 'IN', 'INDEX', 'INFINITY', 'INSERT', 'INTO', 'IS', 'KEYSPACE', 'LIMIT', 'MATERIALIZED', 'MODIFY', 'NAN', 'NORECURSIVE', 'NOT', 'NULL', 'OF', 'ON', 'OR', 'ORDER', 'PRIMARY', 'RENAME', 'REVOKE', 'SCHEMA', 'SELECT', 'SET', 'TABLE', 'TO', 'TOKEN', 'TRUNCATE', 'UNLOGGED', 'UPDATE', 'USE', 'USING', 'VIEW', 'WHERE', 'WITH']
}

module.exports = Constants
